### PR TITLE
100k peers project

### PR DESCRIPTION
**Pull request by rustyrussell**
(Based on #5868 which changes htables to need to be tal'ocated, simply to avoid merge hell!)

I hacked the code to ignore signatures so I could quickly open 100k peers with one channel each, and looked at how the node responded.  Two simple fixes (linked lists were the culprit) and such a node is manageable (though not exactly speedy!).

Before this, reconnecting to all the peers took longer and longer (noticeable before we even had 10k peers).   And startup was extremely slow (not sure how bad it was, I gave up!).

It still takes 9 seconds to listpeers, and almost a minute to startup, and database is about 185MB.

Changelog-None Not user visible.

### CODE REVIEW

**Comment by whitslack** [comment]
> reconnecting to all the peers took longer and longer

Could this be why my `lightning_connectd` eats my CPU for breakfast, lunch, and dinner?

**Comment by rustyrussell** [comment]
> Could this be why my lightning_connectd eats my CPU for breakfast, lunch, and dinner?

Hmm, I don't think so: this was lightningd.  connectd does do all the crypto now, so perhaps it's a bottleneck on large nodes?  The best way to know for sure is:

```
perf record --call-graph dwarf $(pidof lightning_channeld)
```

And let it run for 30 seconds, then ^C and do perf report?

**Comment by rustyrussell** [comment]
Rebased, fixed another test which assumed peer order.

**Comment by cdecker** [comment]
The python linter seems not to be happy with this one:

```
tests/test_pay.py:2419:9: F405 'results' may be undefined, or defined from star imports: fixtures
```

**Comment by rustyrussell** [comment]
"This time for sure!"

File: common/node_id.h [hunk 1]
```diff
@@ -43,4 +46,21 @@ static inline int node_id_idx(const struct node_id *id1,
...
 void fromwire_node_id(const u8 **cursor, size_t *max, struct node_id *id);
+
+/* Hash table functions for node ids */
+static inline const struct node_id *node_id_keyof(const struct node_id *id)
+{
+	return id;
+}
+
+/* We need to define a hashing function. siphash24 is a fast yet
+ * cryptographic hash in ccan/crypto/siphash24; we might be able to get away
```
Comment (cdecker) [inline common/node_id.h:57]: I had to dig into this a bit, it seems the hash function is collision resistant if keyed with a random key, which we do here. It is not generally cryptographically secure, and loses collision resistance if the attacker knows the key.

File: tests/test_pay.py [hunk 2]
```diff
@@ -2413,7 +2413,10 @@ def test_setchannel_all(node_factory, bitcoind):
     wait_for(lambda: [c['base_fee_millisatoshi'] for c in l1.rpc.listchannels(scid3)['channels']] == [0xDEAD, DEF_BASE])
     wait_for(lambda: [c['fee_per_millionth'] for c in l1.rpc.listchannels(scid3)['channels']] == [0xBEEF, DEF_PPM])
 
+    # Don't assume order!
     assert len(result['channels']) == 2
+    if result['channels'][0]['peer_id'] == l3.info['id']:
+        result['channels'] = [result['channels'][1], result['channels'][0]]
```
Comment (cdecker) [inline tests/test_pay.py:2419]:
```
result['channels'] = sorted(result['channels'], key=lambda c: c['peer_id'])
```

File: lightningd/channel_control.c [hunk 3]
```diff
@@ -113,10 +113,11 @@ static void try_update_blockheight(struct lightningd *ld,
 void notify_feerate_change(struct lightningd *ld)
 {
 	struct peer *peer;
+	struct peer_node_id_map_iter it;
 
-	/* FIXME: We should notify onchaind about NORMAL fee change in case
-	 * it's going to generate more txs. */
```
Comment (cdecker) [inline lightningd/channel_control.c:118]: Is this comment no longer necessary?

**Review by cdecker — APPROVED** [review]
That's quite the improvement, nice :-)

ACK https://github.com/ElementsProject/lightning/pull/5849/commits/3672dcd23de2d1502fccd734fd3dd53187ec48b3

File: common/node_id.h (hunk 1 above)
Comment (rustyrussell) [reply common/node_id.h:57]: Indeed!

File: lightningd/channel_control.c (hunk 3 above)
Comment (rustyrussell) [reply lightningd/channel_control.c:118]: Technically still true, but my branch which does zero-fee anchors fixes this (onchaind no longer makes its own txs).  Overzealous removal I guess.

**Comment by rustyrussell** [comment]
Weird, this was already merged by @endothermicdev .  Not sure why GH didn't notice.

**Pull request closed without merging**